packages/contracts-bedrock/test/L2/L2ToL2CrossDomainMessenger.t.sol

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+// Testing utilities
+import { Test } from "forge-std/Test.sol";
+import { Reverter } from "test/mocks/Callers.sol";
+
+// Libraries
+import { Predeploys } from "src/libraries/Predeploys.sol";
+
+// Target contract
+import { CrossL2Inbox } from "src/L2/CrossL2Inbox.sol";
+import { L2ToL2CrossDomainMessenger, NotEntered } from "src/L2/L2ToL2CrossDomainMessenger.sol";
+
+contract L2ToL2CrossDomainMessengerTest is Test {
+    /// @dev L2ToL2CrossDomainMessenger contract instance.
+    L2ToL2CrossDomainMessenger l2ToL2CrossDomainMessenger;
+
+    /// @dev Sets up the test suite.
+    function setUp() public {
+        vm.etch(Predeploys.CROSS_L2_INBOX, address(new CrossL2Inbox()).code);
+        l2ToL2CrossDomainMessenger = new L2ToL2CrossDomainMessenger();
+    }
+
+    /// @dev Tests the `sendMessage` function.
+    function testFuzz_sendMessage_succeeds(
+        uint256 _destination,
+        address _target,
+        bytes memory _message,
+        uint256 _value
+    )
+        external
+    {
+        vm.assume(_destination != block.chainid);
+
+        vm.deal(address(this), _value);
+
+        l2ToL2CrossDomainMessenger.sendMessage{ value: _value }({
+            _destination: _destination,
+            _target: _target,
+            _message: _message
+        });
+    }
+
+    /// @dev Tests that the `sendMessage` function fails when the destination is the same as the source.
+    function test_sendMessage_toSelf_fails() external {
+        vm.expectRevert("L2ToL2CrossDomainMessenger: cannot send message to self");
+        l2ToL2CrossDomainMessenger.sendMessage({
+            _destination: block.chainid,
+            _target: address(0x1234),
+            _message: hex"1234"
+        });
+    }
+
+    /// @dev Tests the `relayMessage` function.
+    function testFuzz_relayMessage_succeeds(
+        uint256 _source,
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        bytes memory _message,
+        uint256 _value
+    )
+        external
+    {
+        vm.assume(_target.code.length == 0);
+
+        vm.mockCall({
+            callee: Predeploys.CROSS_L2_INBOX,
+            data: abi.encodeWithSelector(CrossL2Inbox.origin.selector),
+            returnData: abi.encode(address(l2ToL2CrossDomainMessenger))
+        });
+
+        vm.expectEmit(address(l2ToL2CrossDomainMessenger));
+        emit L2ToL2CrossDomainMessenger.RelayedMessage(
+            keccak256(abi.encode(block.chainid, _source, _nonce, _sender, _target, _message))
+        );
+
```

Comment:
We should have a `vm.expectCall` here too to cover the low level call implementation